docs/architecture/labeling.md

```diff
@@ -0,0 +1,256 @@
+```
+
+#### Getting list of possible labels
+Label definitions are created every time, even when a user directly label Application or Runtime with a new key.
+Thanks to that, to provide a list of possible label keys, we need to return all Label Definition for a specific tenant.
+This functionality can be used in UI, for suggesting already existing label's key.
+
+#### Search
+There are queries for Applications/Runtimes where user can define LabelFilter:
+```graphql
+ applications(filter: [LabelFilter!], first: Int = 100, after: PageCursor):  ApplicationPage!
+```
+
+Because now every label is defined by JSON Schema, LabelFilter needs to be changed, from:
+```graphql
+input LabelFilter {
+    label: String!
+    values: [String!]!
+    operator: FilterOperator = ALL
+}
+```
+to: 
+```graphql
+input LabelFilter {
+    label: String!
+    query: String # optional, if not provided returns every object with given label regardless of its value. 
+}
+```
+
+Challenging part is how the user will provide **query** field.
+There is no standard query language for JSON, see [discussion](https://stackoverflow.com/questions/777455/is-there-a-query-language-for-json).
+We have many alternatives:
+- JSON Path (https://goessner.net/articles/JsonPath/)
+- jq
+- jmespath
+- mongo DB has it's own query language
+- kubectl does not mention which standard they are supporting: https://kubernetes.io/docs/reference/kubectl/jsonpath/
+but it looks the same as Goessner's JSON Path. It looks that they implemented parsing on their own:
+ https://github.com/kubernetes/kubernetes/blob/7faeee22b109d644f9a0ed736447e8867cae728e/staging/src/k8s.io/client-go/util/jsonpath/jsonpath.go
+
+- SQL/JSON Path Expressions (part of SQL-2016 specification http://www.sai.msu.su/~megera/postgres/talks/jsonpath-pgday.it-2019.pdf, https://news.ycombinator.com/item?id=19949240). 
+
+Options enumerated above are not compatible with each other.
+
+The simplest solution will be to use SQL/JSON Path, then we can propagate that value directly to the PostgreSQL. 
+See [this section](#database-schema)to learn how it can be implemented.
+Unfortunately, this functionality is planned for PostgreSQL 12, which is going to be released in Q3 2019, see [roadmap](https://www.postgresql.org/developer/roadmap/) and [features highlights](https://www.postgresql.org/about/news/1943/).
+We don't know when this version will be available on GCP or AWS, so, for now, we will be forced to use Postgres running inside the cluster.
+Also, not all relational databases support JSON Path Expressions, other than Postgres is [SQL Server](https://docs.microsoft.com/en-us/sql/relational-databases/json/json-path-expressions-sql-server?view=sql-server-2017) 
+
+#### Special case: Scenario Label
+For scenario label we have additional requirements:
+
+- there is an always `default` scenario
+- every application/runtime has to be assigned to at least one scenario. If not specified explicitly, the `default` scenario is used.
```

Comment:
I'm not sure if every and each runtime should land in the default scenario. Adding new runtimes will always then already got default configuration.